kingpin/actors/aws/cloudformation.py

```diff
@@ -212,35 +212,50 @@ def _discover_default_params(self, template_body):
         }
         return default_params
 
-    def _get_template_body(self, template):
+    def _get_template_body(self, inline_template, s3_bucket, s3_key, region):
         """Reads in a local template file and returns the contents.
 
         If the template string supplied is a local file resource (has no
         URI prefix), then this method will return the contents of the file.
         Otherwise, returns None.
 
         Args:
-            template: String with a reference to a template location.
 
         Returns:
-            One tuple of:
-              (Contents of template file, None)
-              (None, URL of template)
+          (Contents of template file, None)
+          (Contents of template downloaded from s3, URL of template)
 
         Raises:
             InvalidTemplate
         """
-        if template is None:
-            return None, None
-
-        remote_types = ('http://', 'https://')
-
-        if template.startswith(remote_types):
-            return None, template
+        # if the template is provided inline, return it and don't bother with
+        # s3 buckets
+        if inline_template:
+            try:
+                return json.dumps(self._parse_policy_json(inline_template)), None
+            except exceptions.UnrecoverableActorFailure as e:
+                raise InvalidTemplate(e)
 
         try:
-            return json.dumps(self._parse_policy_json(template)), None
-        except exceptions.UnrecoverableActorFailure as e:
+            # the template is on S3.  Download it so we can inspect it and
+            # construct the proper S3 url.
+            if s3_bucket is None:
+                raise Exception('s3_bucket not specified')
+            if s3_key is None:
+                raise Exception('s3_key not specified')
+
+            # make sure bucket is in correct region
+            resp = self.s3_conn.get_bucket_location(s3_bucket)
+            bucket_region = resp['LocationConstraint']
+
+            if bucket_region != region:
+                raise Exception('bucket region does not match')
+
+            resp = self.api_call(self.s3_conn.get_object, {'Bucket': s3_bucket, 'Key': s3_key})
+            remote_template = resp['Body'].read()
+            url = f'/https://s3.{region}.amazonaws.com/{s3_bucket}/{s3_key}'
```

Comment:
Whats the extra slash at the beginning? Typo?

Reply:
typo